src/persist/src/s3.rs

```diff
@@ -491,19 +492,25 @@ impl Blob for S3Blob {
                 let enable_one_allocation =
                     ENABLE_ONE_ALLOC_PER_REQUEST.get(&self.cfg) && enable_s3_lgalloc;
                 let mut buffer = match object.content_length() {
-                    Some(len @ 1..) if enable_one_allocation => {
-                        let len: u64 = len.try_into().expect("positive integer");
-                        // N.B. `lgalloc` cannot reallocate so we need to make sure the initial
-                        // allocation is large enough to fit then entire blob.
-                        let buf: MetricsRegion<u8> = self
-                            .metrics
-                            .lgbytes
-                            .persist_s3
-                            .new_region(usize::cast_from(len));
-                        Some(buf)
+                    Some(len @ 1..) => {
+                        if enable_one_allocation {
+                            let len: u64 = len.try_into().expect("positive integer");
+                            // N.B. `lgalloc` cannot reallocate so we need to make sure the initial
+                            // allocation is large enough to fit then entire blob.
+                            let buf: MetricsRegion<u8> = self
+                                .metrics
+                                .lgbytes
+                                .persist_s3
+                                .new_region(usize::cast_from(len));
+                            Some(buf)
+                        } else {
+                            None
+                        }
                     }
-                    Some(len) => {
-                        tracing::error!(?len, "found invalid content-length, falling back");
+                    // content-length of 0 isn't necessarily invalid.
+                    Some(len @ ..=-1) => {
+                        tracing::trace!(?len, "found invalid content-length, falling back");
+                        get_invalid_resp.inc();
                         None
                     }
                     _ => None,
```

Comment:
Is rust smart enough yet to let you replace the `_` with `Some(0) | None`?

Reply:
turns out it is 🙌